arch: riscv: core: Place vectors section through zephyr_linker_sources()

msierszulski opened the pull request:
This PR is fixing placing the vectors section through `zephyr_linker_sources(ROM_START ...)` (as done in the ARM architecture port) so its order can be adjusted by `SORT_KEY`.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49903

msierszulski committed:
Revert "tests: kernel: gen_isr_table: Disable RISC-V direct ISR test"

This reverts commit 857f42570c79feff4a9673c0557daef5aa98f896.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
arch: riscv: core: Place vectors section through zephyr_linker_sources()

This commit is fixing placing the vectors section through
zephyr_linker_sources(ROM_START ...) (as done in the ARM
architecture port) so its order can be adjusted by SORT_KEY.

Fixes #49903

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski commented:
cc @stephanosio

stephanosio reviewed (changes requested):
Unfortunately, the fix is not going to be as simple as this -- i.e. we need to make sure it works for the rest of the SoCs thanks to how they all do their own thing ...

msierszulski committed:
soc: riscv: ae350: Remove redundant .vectors sections

This commit removes doubled .vectors sections in ae350 SoC.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski commented:
@stephanosio  I have followed your comments. Can you look at the code now?

stephanosio commented on soc/riscv/riscv-privilege/telink_b91/CMakeLists.txt:
```diff
@@ -13,4 +13,3 @@ zephyr_ld_options(-fuse-ld=bfd)
 # Set compile options
 zephyr_compile_options_ifdef(CONFIG_TELINK_B91_HWDSP -mext-dsp)
 zephyr_compile_options_ifndef(CONFIG_RISCV_GP -mno-relax)
-zephyr_linker_sources(ROM_START init.ld)
```
With this change, the `.init.*` sections are now being discarded for this platform (see the map file).

msierszulski replied:
I changed it. Added `SORT_KEY 0x0` and now `.init.*` sections come before the vector sections.

stephanosio replied:
Confirmed, thanks.

msierszulski committed:
soc: riscv: telink_b91: Place .init before .vectors section

This commit fixes placing .init sections before .vectors
sections in telink_b91 SoC.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

stephanosio commented on include/zephyr/arch/riscv/common/linker.ld:
```diff
@@ -119,13 +119,6 @@ SECTIONS
 #include <snippets-rom-start.ld>
     } GROUP_LINK_IN(ROMABLE_REGION)
 
-    SECTION_PROLOGUE(_VECTOR_SECTION_NAME,,)
```
should remove `_VECTOR_SECTION_NAME` definition at L30 since it is no longer being used

msierszulski replied:
Sure, thanks!

stephanosio commented on soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld:
```diff
@@ -116,18 +116,10 @@ SECTIONS
     GROUP_START(ROMABLE_REGION)
     __rom_region_start = ROM_BASE;
 
-    SECTION_PROLOGUE(_VECTOR_SECTION_NAME,,)
```
should remove `_VECTOR_SECTION_NAME` definition at L31 since it is no longer being used

msierszulski replied:
Sure, thanks!

stephanosio reviewed (commented):
Looks ok in general. Just some final clean-up related comments.

stephanosio reviewed (approved):
LGTM
